File: apps/android/app/src/main/java/com/signalasi/chat/AgentCodeDiscussionPolicy.kt
package com.signalasi.chat

import java.util.Locale

internal object AgentCodeDiscussionPolicy {
    fun isInformational(goal: String): Boolean {
        val normalized = AgentUntrustedEvidenceBoundary.trustedInstructionPrefix(goal)
            .lowercase(Locale.US)
            .replace(Regex("\\s+"), " ")
            .trim()
        return DISCUSSION_PATTERN.containsMatchIn(normalized) &&
            !EXECUTION_OVERRIDE_PATTERN.containsMatchIn(normalized)
    }

    private val DISCUSSION_PATTERN = Regex(
        "(?:\\b(?:explain|describe|compare|summarize|discuss)\\b|" +
            "(?:\u8bf4\u660e|\u89e3\u91ca|\u63cf\u8ff0|\u6bd4\u8f83|\u603b\u7ed3|\u8ba8\u8bba)|" +
            "\\b(?:list|outline|describe|suggest|propose|explain)\\b.{0,96}" +
            "\\b(?:test cases?|unit tests?|test scenarios?)\\b|" +
            "\\b(?:what happens|what occurs|why|how|explain|describe)\\b.{0,128}" +
            "\\b(?:code|function|async|await|promise|error|exception|bug|algorithm)\\b|" +
            "\\b(?:give|provide|show)\\b.{0,64}" +
            "\\b(?:example|sample|snippet|pseudocode|fix|approach)\\b|" +
            "(?:\u5217\u51fa|\u7ed9\u51fa|\u8bf4\u660e|\u63cf\u8ff0|\u5efa\u8bae|\u8bbe\u8ba1).{0,64}" +
            "(?:\u5355\u5143\u6d4b\u8bd5|\u6d4b\u8bd5\u573a\u666f|\u6d4b\u8bd5\u7528\u4f8b)|" +
            "(?:\u4f1a\u9020\u6210\u4ec0\u4e48|\u4f1a\u53d1\u751f\u4ec0\u4e48|\u4e3a\u4ec0\u4e48|\u4e3a\u4f55|" +
            "\u5982\u4f55|\u89e3\u91ca|\u8bf4\u660e).{0,96}" +
            "(?:\u4ee3\u7801|\u51fd\u6570|\u5f02\u6b65|\u9519\u8bef|\u5f02\u5e38|\u7b97\u6cd5|bug|await|promise)|" +
            "(?:\u7ed9\u51fa|\u63d0\u4f9b|\u5c55\u793a).{0,64}" +
            "(?:\u793a\u4f8b|\u4f8b\u5b50|\u6837\u4f8b|\u4ee3\u7801\u7247\u6bb5|\u4f2a\u4ee3\u7801|\u4fee\u590d\u601d\u8def|\u4fee\u590d\u65b9\u6848))",
        RegexOption.IGNORE_CASE
    )

    private val EXECUTION_OVERRIDE_PATTERN = Regex(
        "(?:\\b(?:analyze|inspect|review|audit)\\b.{0,64}" +
            "\\b(?:project|repository|repo|codebase|files?)\\b|" +
            "(?:\u5206\u6790|\u68c0\u67e5|\u5ba1\u67e5|\u5ba1\u8ba1).{0,64}" +
            "(?:\u9879\u76ee|\u4ed3\u5e93|\u4ee3\u7801\u5e93|\u6587\u4ef6)|" +
            "\\b(?:write|create|implement|run|execute|add|modify|edit|fix)\\b.{0,96}" +
            "\\b(?:tests?|unit tests?|test cases?|code|function|program|project|repository|files?|" +
            "bugs?|errors?|exceptions?)\\b|" +
            "\\b(?:write|create|run|execute)\\b.{0,64}\\b(?:examples?|samples?)\\b|" +
            "(?:\u7f16\u5199|\u521b\u5efa|\u5b9e\u73b0|\u8fd0\u884c|\u6267\u884c|\u6dfb\u52a0|" +
            "\u4fee\u6539|\u7f16\u8f91|\u4fee\u590d).{0,64}" +
            "(?:\u6d4b\u8bd5|\u4ee3\u7801|\u51fd\u6570|\u7a0b\u5e8f|\u9879\u76ee|\u4ed3\u5e93|\u6587\u4ef6|" +
            "\u9519\u8bef|\u5f02\u5e38|\u95ee\u9898|\u5b83\u4eec|\u8fd9\u4e9b)|" +
            "(?:\u7f16\u5199|\u521b\u5efa|\u8fd0\u884c|\u6267\u884c).{0,64}(?:\u793a\u4f8b|\u4f8b\u5b50))",
        RegexOption.IGNORE_CASE
    )
}

File: apps/ios/SignalASI/AgentCodeDiscussionPolicy.swift
import Foundation

enum AgentCodeDiscussionPolicy {
  static func isInformational(_ goal: String) -> Bool {
    let normalized = AgentUntrustedEvidenceBoundary.trustedInstructionPrefix(goal)
      .lowercased(with: Locale(identifier: "en_US_POSIX"))
      .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
      .trimmingCharacters(in: .whitespacesAndNewlines)
    return matches(discussionPattern, in: normalized) &&
      !matches(executionOverridePattern, in: normalized)
  }

  private static func matches(_ pattern: String, in value: String) -> Bool {
    value.range(of: pattern, options: [.regularExpression, .caseInsensitive]) != nil
  }

  private static let discussionPattern =
    "(?:\\b(?:list|outline|describe|suggest|propose|explain)\\b.{0,96}" +
    "\\b(?:test cases?|unit tests?|test scenarios?)\\b|" +
    "(?:\u{5217}\u{51fa}|\u{7ed9}\u{51fa}|\u{8bf4}\u{660e}|\u{63cf}\u{8ff0}|\u{5efa}\u{8bae}|" +
    "\u{8bbe}\u{8ba1}).{0,64}(?:\u{5355}\u{5143}\u{6d4b}\u{8bd5}|\u{6d4b}\u{8bd5}\u{573a}\u{666f}|" +
    "\u{6d4b}\u{8bd5}\u{7528}\u{4f8b}))"

  private static let executionOverridePattern =
    "(?:\\b(?:write|create|implement|run|execute|add|modify|edit|fix)\\b.{0,96}" +
    "\\b(?:tests?|unit tests?|test cases?|code|function|program|project|repository|files?)\\b|" +
    "(?:\u{7f16}\u{5199}|\u{521b}\u{5efa}|\u{5b9e}\u{73b0}|\u{8fd0}\u{884c}|\u{6267}\u{884c}|" +
    "\u{6dfb}\u{52a0}|\u{4fee}\u{6539}|\u{7f16}\u{8f91}|\u{4fee}\u{590d}).{0,64}" +
    "(?:\u{6d4b}\u{8bd5}|\u{4ee3}\u{7801}|\u{51fd}\u{6570}|\u{7a0b}\u{5e8f}|\u{9879}\u{76ee}|" +
    "\u{4ed3}\u{5e93}|\u{6587}\u{4ef6}|\u{5b83}\u{4eec}|\u{8fd9}\u{4e9b}))"
}
